README.md

```diff
@@ -43,45 +43,94 @@ FormatParser.parse(File.open("myimage", "rb"), natures: [:video, :image], format
+your method or `break` your Proc as early as possible. A blank `return` works fine too.
 
-1) Instances of the new parser class needs to respond to a `call` method which takes one IO object as an argument and returns some metadata information about its corresponding file or nil otherwise.
-2) Instances of the new parser class needs to respond `natures` and `formats` accessor methods, both returning an array of symbols. A simple DSL is provided to avoid writing those accessors.
-3) The class needs to register itself as a parser.
+The IO will at the minimum support the subset of the IO API defined in `ConstrainedIO`
 
+Strictly, a parser should be one of the two things:
+
+1) An object that can be `call()`-ed itself, with an argument that conforms to `ConstrainedIO`
+2) An object that responsds to `new` and returns something that can be `call()`-ed with the same convention.
+
+The second opton is useful for parsers that are stateful and non-reentrant. FormatParser is made to be used in
+threaded environments, and if you use instance variables you need your parser to be isolated from it's siblings in
+other threads - therefore you can pass a Class on registration to have your parser instantiated for each `call()`,
+anew.
+
+Your parser has to be registered using `FormatParser.register_parser` with the information on the formats
+and file natures it provides.
 
 Down below you can find a basic parser implementation:
 
 ```ruby
-class BasicParser
-  include FormatParser::DSL # Adds formats and natures methods to the class, which define
-                            # accessor for all the instances.
-  
-  formats :foo, :baz # Indicates which formats it can read.
-  natures :bar       # Indicates which type of file from a human perspective it can read:
-                     #      - :audio
-                     #      - :document
-                     #      - :image
-                     #      - :video
-  def call(file)
-    # Returns a DTO object with including some metadata.
+MyParser = ->(io) {
+  # Make sure you are only using the right subset of IO methods
+  io = FormatParser::ConstrainedIO.new(io)
+  # ... do some parsing with `io`
+  magic_bytes = io.read(4)
+  break if magic_bytes != 'XBMP'
+  # ... more parsing code
+  # ...and return the FileInformation::Image object with the metadata.
+  FormatParser::Image.new(
+    width_px: parsed_width,
+    height_px: parsed_height,
+  )
+}
+
+# Register the parser with the module, so that it will be applied to any
+# document given to `FormatParser.parse()`. The supported natures are currently
+#      - :audio
+#      - :document
+#      - :image
+#      - :video
+FormatParser.register_parser MyParser, natures: :image, formats: :bmp
+```
+
+If you are using a class, this is the skeleton to use:
+
+```ruby
+class MyParser
+  def call(io)
+    # Make sure you are only using the right subset of IO methods
+    io = FormatParser::ConstrainedIO.new(io)
```

Comment:
Hm. Hmmm.

I think this is a code smell. To reassign an argument makes me think whether the parameter passing in the method invocation was imprecise (maybe a ConstrainedIO has to be enforced/assumed) or that you need a need var. I honestly dislike to modify arguments or reassigning variables. Just my two cents though. 🙃

Reply:
I do it all the time primarily because we don't need to have strictly-typed references in Ruby, but I do get the idea that suggesting this in a README might be a bad idea education-wise ;-)